### PR TITLE
Delete all live docs when query matched a whole segment.

### PR DESCRIPTION
**Pull request by vsop-479**
### Description

We can delete all live docs when a query matched a whole segment, in `FrozenBufferedUpdates.applyQueryDeletes`.

I have implemented it for `PointRangeQuery`, and also trying to implement it for other queries.

### CODE REVIEW

**Comment by vsop-479** [comment]
@mikemccand 
Please take a look when you get a chance.

**Review by mikemccand — COMMENTED** [review]
I am nervous about this change -- altering the very widely used `Scorer` API to express whether it matches all documents feels invasive.

Also, the use case this is optimizing for, feels rare?  Though, perhaps if an index sorts its segments by time range, and collates documents into segments by time range as well, delete by range query over a time range might more efficiently drop whole segments if we could do something like this cleanly?  Even so, it's quite rare that an index would invoke this pruning, and the likely smallish time it takes today is fine?

File: lucene/core/src/java/org/apache/lucene/search/Scorer.java [hunk 1]
```diff
@@ -98,4 +100,14 @@ public int advanceShallow(int target) throws IOException {
    * {@link #advanceShallow(int) shallow-advanced} to included and {@code upTo} included.
    */
   public abstract float getMaxScore(int upTo) throws IOException;
+
+  /** Set whether we can match all docs in this scored segment. */
+  public void setMatchAll(boolean isMatchAll) {
```
[the remark this answers is not in the document]
Comment (vsop-479) [reply lucene/core/src/java/org/apache/lucene/search/Scorer.java:105]: >  it means consumers of Scorer can suddenly setMatchAll without being true?

Yes, I should found a better way to mark `isMatchAll`.

**Comment by vsop-479** [comment]
> Also, the use case this is optimizing for, feels rare?

> Even so, it's quite rare that an index would invoke this pruning, and the likely smallish time it takes today is fine?

Yes, delete by range query over a time range, is the only case which came to my mind, but the improve maybe will not be significant.

**Comment by vigyasharma** [comment]
> I am nervous about this change -- altering the very widely used `Scorer` API to express whether it matches all documents feels invasive.

+1. I wonder how widely useful a `matchesAll()` API in the scorer would be. The only place that comes to mind is dropping a segment with all deleted docs (i.e. this PR). For search scenarios, we'd need to iterate over the scorer iterator; and the `cost()` API is good enough for estimation use cases.

**Comment by vsop-479** [comment]
Thanks for reviewing @vigyasharma . I will close this PR.

**Pull request closed without merging**